utils/near-performance-metrics/src/framed_write.rs

```diff
@@ -0,0 +1,369 @@
+use std::cell::RefCell;
+use std::marker::PhantomData;
+use std::ops::DerefMut;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::task::{Context, Poll};
+use std::{io, task};
+
+use actix::{Actor, ActorContext, ActorFuture, AsyncContext, Running, SpawnHandle};
+use bitflags::bitflags;
+use bytes::BytesMut;
+use tokio::io::{AsyncWrite, AsyncWriteExt};
+use tokio_util::codec::Encoder;
+
+// This file was copied from actix-0.11.0-beta.1/src/io.rs
+// The only change was to add EncoderCallBack
+
+bitflags! {
+    struct Flags: u8 {
+        const CLOSING = 0b0000_0001;
+        const CLOSED = 0b0000_0010;
+    }
+}
+
+const LOW_WATERMARK: usize = 4 * 1024;
+const HIGH_WATERMARK: usize = 4 * LOW_WATERMARK;
+
+#[allow(unused_variables)]
+pub trait WriteHandler<E>
+where
+    Self: Actor,
+    Self::Context: ActorContext,
+{
+    /// Called when the writer emits error.
+    ///
+    /// If this method returns `ErrorAction::Continue` writer processing
+    /// continues otherwise stream processing stops.
+    fn error(&mut self, err: E, ctx: &mut Self::Context) -> Running {
+        Running::Stop
+    }
+
+    /// Called when the writer finishes.
+    ///
+    /// By default this method stops actor's `Context`.
+    fn finished(&mut self, ctx: &mut Self::Context) {
+        ctx.stop()
+    }
+}
+
+struct InnerWriter<E: From<io::Error>, K: EncoderCallBack> {
+    flags: Flags,
+    buffer: BytesMut,
+    error: Option<E>,
+    low: usize,
+    high: usize,
+    handle: SpawnHandle,
+    task: Option<task::Waker>,
+    callback: K,
+}
+
+pub struct EmptyCallBack {}
+
+impl EncoderCallBack for EmptyCallBack {
+    fn drained(&mut self, _bytes: usize, _buf_len: usize, _buf_capacity: usize) {}
+}
+
+struct UnsafeWriter<T: AsyncWrite, E: From<io::Error>, K: EncoderCallBack>(
+    Rc<RefCell<InnerWriter<E, K>>>,
+    Rc<RefCell<T>>,
+);
+
+impl<T: AsyncWrite, E: From<io::Error>, K: EncoderCallBack> Clone for UnsafeWriter<T, E, K> {
+    fn clone(&self) -> Self {
+        UnsafeWriter(self.0.clone(), self.1.clone())
+    }
+}
+
+pub trait EncoderCallBack {
+    fn drained(&mut self, bytes: usize, buf_len: usize, buf_capacity: usize);
```

Comment:
what is this supposed to do

Reply:
I added a new callback function, which is given as parameter to `FramedWriter`. Whenever buffer decreased in size, we get information about the number of bytes consumed, and size of Buffer, and capacity of the butter.

We also have `Encoder` trait, which get called whenever new data gets added to the buffer. So, we know size/capacity of the buffer, as well as the number of bytes added to the buffer during that time. 

With both `Encoder` and `EncoderCallBack` traits you have enough information to know the current state of the buffer, there is no need to have login inside `FramedWrite`.